keps/0017-pipe-tasks.md

```diff
@@ -0,0 +1,122 @@
+
+## Motivation
+
+We want to ease the development of operators with complicated life-cycles by enabling to pipe files from one task to another. While this can be achieved in a manual way, involving custom scripts, this is exactly the kind of "magic glue" that KUDO could provide to the operator developers to ease the development process.
+
+### Goals
+
+The goal of this KEP is to describe application examples, limitations and ways to implement passing files between tasks. We aim to help operator developers pipe small files (<1Mb) like certificates and dynamically generated configuration which can be referenced later as ConfigMaps and Secrets.
+
+### Non-Goals
+
+Allowing to pipe all kind of files (>1Mb) between tasks requires a general-purpose storage solution (e.g. S3) available in the cluster. While certainly possible, this is not the goal of this KEP.
+
+## Proposal
+
+This section describes how pipe tasks and files they produce are configured in the operator. Here is a task definition that produces a file that will be stored as a Secret, referenced by `{{.Pipes.Certificate}}` key:
+```yaml
+steps:
+    - name: bootstrap
+    tasks:
+        - gencert
+            - pipe:
+                - file: /usr/share/MyKey.key
+                  kind: Secret # ConfigMap
+                  key: {{.Pipes.Certificate}}
+``` 
+The key `{{.Pipes.Certificate}}` value will be generated by KUDO to avoid collisions and to make sure that it is not used before the file is generated. In the above example we would generate a secret name like `deploy.bootstrap.gencert.pipes.certificate-#hash` to capture which plan/phase/step/task it belongs too as well as the hash of its content. KUDO would use this hash to avoid recreating secrets for unchanged files on a plan rerun. We would also use labels ensuring that the secret is cleaned up when the corresponding Instance is removed.
+
+The corresponding `gencert` task can only have one resource definition:
+```yaml
+tasks:
+  gencert:
+    resources:
+      - gencert.yaml
+```
+
+For the pipe task resources, we only allow a [ContainerSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#container-v1-core) to be specified. Reasons for that are explained below in the implementation details. The ContainerSpec has to define a shared volume where the files are stored.
+
+```yaml
+spec:
+  volumes:
+  - name: shared-data
+    emptyDir: {}
+
+  containers:
+    - name: gencert
+      image: frapsoft/openssl
+      imagePullPolicy: Always
+      command: [ "sh", "-c" ]
+      args:
+        - openssl req -new -newkey rsa:4096 -x509 -sha256 -days 365 -nodes -out MyCertificate.crt -keyout /usr/share/MyKey.key
+      volumeMounts:
+        - name: shared-data
+          mountPath: /usr/share/
+```
+
+Any subsequent task resource might reference previously generated file by its key `{{.Pipes.Certificate}}` e.g.:
```

Comment:
please say "subsequent step" not task and maybe even note that the step has to be ensured to run after the step generating the pipe output just for the sake of clarity :)